File: src/seshat/engine.py
"""The pattern engine — runs detectors over normalized source, honestly.

Input is **normalized** source (comments/strings stripped, deps dropped — see
:mod:`seshat.normalize`), so brace-balancing and line numbers are reliable.

Matcher mechanics (one unified evaluator behind the four declared kinds):
- ``scope: file`` — match per line over the whole source (default).
- ``scope: function`` — match within each brace-balanced function body, with
  ``requires``/``forbids`` evaluated in that function's text. Used for
  "function with/without modifier" (``function_signature`` / ``ast``) rules.
- ``multiline: true`` — match across the whole source (line taken from the match
  start), for cross-line shapes.
- ``requires`` (all must appear in scope) / ``forbids`` (none may) compose
  detectors into semantic checks without a real parser.

Two honesty contracts: a detector below the **confidence threshold** is skipped,
and findings are **deduped per (pattern, line)**.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from . import normalize
from .patterns import Detector, Pattern

# `receive`/`fallback` are functions too, but carry no `function` keyword.
_FUNCTION_RE = re.compile(r"\b(?:function|receive|fallback)\b")


@dataclass
class EngineFinding:
    pattern_id: str
    pattern_name: str
    severity: str
    confidence: float
    category: str
    line: int
    snippet: str
    description: str
    recommendation: str
    swc: str = ""
    cwe: str = ""


def iter_function_regions(text: str) -> list[tuple[str, int]]:
    """Yield ``(function_text, base_line)`` for each function body in ``text``.

    Relies on normalized source (no comments/strings) so brace counting is safe.
    Function *declarations* (no body, terminated by ``;``) are skipped.
    """
    regions: list[tuple[str, int]] = []
    n = len(text)
    for m in _FUNCTION_RE.finditer(text):
        j = m.end()
        depth_paren = 0
        brace = -1
        while j < n:
            ch = text[j]
            if ch == "(":
                depth_paren += 1
            elif ch == ")":
                depth_paren -= 1
            elif ch == ";" and depth_paren <= 0:
                break  # declaration only, no body
            elif ch == "{" and depth_paren <= 0:
                brace = j
                break
            j += 1
        if brace == -1:
            continue
        depth = 0
        k = brace
        while k < n:
            if text[k] == "{":
                depth += 1
            elif text[k] == "}":
                depth -= 1
                if depth == 0:
                    break
            k += 1
        end = min(k + 1, n)
        region = text[m.start():end]
        base_line = text[: m.start()].count("\n") + 1
        regions.append((region, base_line))
    return regions


def _all_present(regexes, text) -> bool:
    return all(rx.search(text) for rx in regexes)


def _any_present(regexes, text) -> bool:
    return any(rx.search(text) for rx in regexes)


def _detector_hits(det: Detector, text: str) -> list[int]:
    """Return the line numbers (1-based) where ``det`` fires in ``text``."""
    if det._rx is None:
        det.compile()

    if det.scope == "function":
        hits: list[int] = []
        for region, base_line in iter_function_regions(text):
            if det._req and not _all_present(det._req, region):
                continue
            if det._forb and _any_present(det._forb, region):
                continue
            # flag the function once, at the first matching line
            for idx, line in enumerate(region.split("\n")):
                if det._rx.search(line):
                    hits.append(base_line + idx)
                    break
        return hits

    # file scope: requires/forbids evaluated over the whole source
    if det._req and not _all_present(det._req, text):
        return []
    if det._forb and _any_present(det._forb, text):
        return []

    if det.multiline:
        return [text[: m.start()].count("\n") + 1 for m in det._rx.finditer(text)]

    hits = []
    for idx, line in enumerate(text.split("\n")):
        if det._rx.search(line):
            hits.append(idx + 1)
    return hits


def scan_source(
    text: str,
    patterns: list[Pattern],
    *,
    min_confidence: float = 0.0,
) -> list[EngineFinding]:
    """Scan one normalized source blob, returning deduped findings.

    Dedup is per ``(pattern_id, line)``; the highest-confidence detector wins.
    """
    lines = text.split("\n")
    best: dict[tuple[str, int], EngineFinding] = {}

    for pat in patterns:
        for det in pat.detectors:
            if det.confidence < min_confidence:
                continue
            for line_no in _detector_hits(det, text):
                key = (pat.id, line_no)
                snippet = lines[line_no - 1].strip()[:160] if 0 < line_no <= len(lines) else ""
                cand = EngineFinding(
                    pattern_id=pat.id,
                    pattern_name=pat.name,
                    severity=pat.severity,
                    confidence=det.confidence,
                    category=pat.category,
                    line=line_no,
                    snippet=snippet,
                    description=det.description,
                    recommendation=det.recommendation,
                    swc=pat.swc,
                    cwe=pat.cwe,
                )
                prev = best.get(key)
                if prev is None or cand.confidence > prev.confidence:
                    best[key] = cand

    return sorted(best.values(), key=lambda f: (f.line, f.pattern_id))


def scan_snippet(src: str, pattern: Pattern) -> list[EngineFinding]:
    """Normalize a raw Solidity snippet and scan it with a single pattern."""
    norm = normalize.normalize_text(src)
    return scan_source(norm.content, [pattern], min_confidence=0.0)


def check_pattern_fixtures(pattern: Pattern) -> list[str]:
    """Verify a pattern's inline fixtures: positives fire, negatives stay quiet.

    Returns a list of failure messages (empty ⇒ the pattern proves itself).
    """
    failures: list[str] = []
    for i, src in enumerate(pattern.positive):
        hits = [f for f in scan_snippet(src, pattern) if f.pattern_id == pattern.id]
        if not hits:
            failures.append(f"{pattern.id}: positive fixture #{i + 1} did not fire")
    for i, src in enumerate(pattern.negative):
        hits = [f for f in scan_snippet(src, pattern) if f.pattern_id == pattern.id]
        if hits:
            lines = sorted({h.line for h in hits})
            failures.append(
                f"{pattern.id}: negative fixture #{i + 1} false-positived at lines {lines}"
            )
    return failures

File: rust/src/engine.rs
//! The matcher — a faithful port of `seshat.engine`.
//!
//! Same mechanics, so findings are identical to the Python engine:
//! - `scope=file`  → per-line match over the whole source
//! - `scope=function` → match inside each brace-balanced function body
//!   (including `receive`/`fallback`); flag once, at the first matching line
//! - `multiline` → match over the whole source, line taken from match start
//! - `requires` (all present in scope) / `forbids` (none present)
//! - confidence threshold + per-(pattern, line) dedup, highest confidence wins

use std::collections::HashMap;

use fancy_regex::Regex;

use crate::pattern::{Detector, Pattern};

#[derive(Clone, Debug, PartialEq)]
pub struct Finding {
    pub pattern_id: String,
    pub severity: String,
    pub confidence: f64,
    pub line: usize,
}

fn rx_match(rx: &Regex, hay: &str) -> bool {
    rx.is_match(hay).unwrap_or(false)
}

fn all_present(rxs: &[Regex], text: &str) -> bool {
    rxs.iter().all(|r| rx_match(r, text))
}

fn any_present(rxs: &[Regex], text: &str) -> bool {
    rxs.iter().any(|r| rx_match(r, text))
}

/// `(function_text, base_line)` for each function body. Mirrors
/// `iter_function_regions` in Python (keyword incl. receive/fallback, brace
/// balanced, declarations skipped).
fn function_regions(text: &str) -> Vec<(&str, usize)> {
    let kw = Regex::new(r"\b(?:function|receive|fallback)\b").unwrap();
    let bytes = text.as_bytes();
    let n = bytes.len();
    let mut regions = Vec::new();

    let mut search_from = 0usize;
    while let Ok(Some(m)) = kw.find_from_pos(text, search_from) {
        let start = m.start();
        search_from = m.end();

        // find the body-opening brace, skipping the signature's parens
        let mut j = m.end();
        let mut depth_paren = 0i32;
        let mut brace: Option<usize> = None;
        while j < n {
            match bytes[j] {
                b'(' => depth_paren += 1,
                b')' => depth_paren -= 1,
                b';' if depth_paren <= 0 => break, // declaration, no body
                b'{' if depth_paren <= 0 => {
                    brace = Some(j);
                    break;
                }
                _ => {}
            }
            j += 1;
        }
        let bstart = match brace {
            Some(b) => b,
            None => continue,
        };

        // balance braces
        let mut depth = 0i32;
        let mut k = bstart;
        let mut end = n;
        while k < n {
            match bytes[k] {
                b'{' => depth += 1,
                b'}' => {
                    depth -= 1;
                    if depth == 0 {
                        end = k + 1;
                        break;
                    }
                }
                _ => {}
            }
            k += 1;
        }
        let base_line = text[..start].matches('\n').count() + 1;
        regions.push((&text[start..end], base_line));
    }
    regions
}

fn detector_hits(det: &Detector, text: &str) -> Vec<usize> {
    if det.scope == "function" {
        let mut hits = Vec::new();
        for (region, base_line) in function_regions(text) {
            if !det.req.is_empty() && !all_present(&det.req, region) {
                continue;
            }
            if !det.forb.is_empty() && any_present(&det.forb, region) {
                continue;
            }
            for (idx, line) in region.split('\n').enumerate() {
                if rx_match(&det.rx, line) {
                    hits.push(base_line + idx);
                    break; // flag the function once
                }
            }
        }
        return hits;
    }

    // file scope: requires/forbids over the whole source
    if !det.req.is_empty() && !all_present(&det.req, text) {
        return Vec::new();
    }
    if !det.forb.is_empty() && any_present(&det.forb, text) {
        return Vec::new();
    }

    if det.multiline {
        let mut hits = Vec::new();
        let mut from = 0usize;
        while let Ok(Some(m)) = det.rx.find_from_pos(text, from) {
            hits.push(text[..m.start()].matches('\n').count() + 1);
            from = if m.end() > m.start() { m.end() } else { m.end() + 1 };
            if from > text.len() {
                break;
            }
        }
        return hits;
    }

    let mut hits = Vec::new();
    for (idx, line) in text.split('\n').enumerate() {
        if rx_match(&det.rx, line) {
            hits.push(idx + 1);
        }
    }
    hits
}

/// Scan normalized source, returning deduped findings sorted by (line, id).
pub fn scan_source(text: &str, patterns: &[Pattern], min_confidence: f64) -> Vec<Finding> {
    let mut best: HashMap<(String, usize), Finding> = HashMap::new();
    for pat in patterns {
        for det in &pat.detectors {
            if det.confidence < min_confidence {
                continue;
            }
            for line in detector_hits(det, text) {
                let key = (pat.id.clone(), line);
                let cand = Finding {
                    pattern_id: pat.id.clone(),
                    severity: pat.severity.clone(),
                    confidence: det.confidence,
                    line,
                };
                match best.get(&key) {
                    Some(prev) if prev.confidence >= cand.confidence => {}
                    _ => {
                        best.insert(key, cand);
                    }
                }
            }
        }
    }
    let mut out: Vec<Finding> = best.into_values().collect();
    out.sort_by(|a, b| a.line.cmp(&b.line).then(a.pattern_id.cmp(&b.pattern_id)));
    out
}
